cps-global-lib/README.md

```diff
@@ -59,28 +68,133 @@ Thereafter it should suffice to run:
 At the base level, any valid Groovy code is OK. So you can define data structures,
 utility functions, and etc., like this:
 
-    $ cat src/org/foo/Point.groovy
-    package org.foo;
+```groovy
+// src/org/foo/Point.groovy
+package org.foo;
 
-    // point in 3D space
-    class Point {
-      float x,y,z;
-    }
+// point in 3D space
+class Point {
+  float x,y,z;
+}
+```
 
 However classes written like this cannot call step functions like `sh` or `git`.
 More often than not, what you want to define is a series of functions that in turn invoke
 other workflow step functions. You can do this by not explicitly defining the enclosing class,
 just like your main workflow script itself:
 
-    $ cat src/org/foo/Zot.groovy
-    package org.foo;
+```groovy
+// src/org/foo/Zot.groovy
+package org.foo;
 
-    def checkOutFrom(repo) {
-      git url: "git@github.com:jenkinsci/${repo}"
-    }
+def checkOutFrom(repo) {
+  git url: "git@github.com:jenkinsci/${repo}"
+}
+```
 
 You can then call such function from your main workflow script like this:
 
-    def z = new org.foo.Zot()
-    z.checkOutFrom(repo)
+```groovy
+def z = new org.foo.Zot()
+z.checkOutFrom(repo)
+```
+
+### Defining global functions
+You can define your own functions that looks and feels like built-in step functions like `sh` or `git`.
+For example, to define `helloWorld` step of your own, create a file named `vars/helloWorld.groovy` and
+define the `call` method:
+
+```groovy
+// vars/helloWorld.groovy
+def call(msg) {
+    // you can call any valid step functions from your code, just like you can from workflow scripts
+    echo "Hello world, ${name}"
+}
+```
+
+Then your workflow can call this function like this:
+
+```groovy
+helloWorld "Joe"
+helloWorld("Joe")
+```
+
+If called with a block, the `call` method will receive a `Closure` object. You can define that explicitly
+as the type to clarify your intent, like the following:
+
+```groovy
+// vars/windows.groovy
+def call(Closure body) {
+    node('windows') {
+        body()
+    }
+}
+```
+
+Your workflow can call this function like this:
+
+```groovy
+windows {
+    bat "cmd /?"
+}
+```
+
+See [the closure chapter of Groovy language reference](http://www.groovy-lang.org/closures.html) for more details
+about the block syntax in Groovy.
+
+### Defining global variables
+Internally, scripts in the `vars` directory are instantiated as a singleton on-demand, when used first.
```

Comment:
(OT:)  at scale (and even not at scale) you likely want code review (e.g. Gerrit / GitHub PR) and may have multiple groups using a single Jenkins instance - so a global cps-global-lib becomes less useful and the desire to store these elsewhere becomes greater.